### ISSUE TITLE
Cuckoo VM Management & Performance Issues (VirtualBox & vSphere)

### ISSUE THREAD

**Issue by MetallDoktor** [cuckoosandbox/cuckoo#2902]
Greetings!

**Issue(s)**

There are some problems on hypervisors we tested with Cuckoo:

VirtualBox

As far as we can see, there is a VM condition cycle in VirtualBox during Cuckoo task execution. Roughly it can be shown like this:

---> Saved ---> Restoring... ---> Running --->
|                                                                        |
<- Reverting <- Powered Off <- Stopping <---

The problem is when several (we ran our tests on ~100 files on 10 VMs) tasks execute simultaneously, VMs change their condition strictly consecutively. For example, until VM1 doesn't change from "Saved" to "Running", VM2 (and then VM3, etc) stay in "Saved". It won't restore a snapshot, it won't start an analysis.

Next, until all VMs specified in hypervisor config reach "Running" state, there is no task timeout check. Because of this, task VM analysis is not limited by its timeout, and task status remains "running" until all VMs change their state (and it is often longer than timeout we set).

Again, all the above is what we can see in VM VirtualBox Manager on Cuckoo Host.

vSphere

In that case, Cuckoo VMs are managed through vCenter, Cuckoo Host locates in the same subnet, but it is a physical server.

There is a long time delay before reverting and powering off VM (log example is shown below). This results in longer overall task execution as well.

**Our configuration**

Cuckoo version: 2.0.7
Host OS: Ubuntu 17.04 Zesty

**Logs and configs**

[conf.zip](https://github.com/cuckoosandbox/cuckoo/files/3846450/conf.zip)

Log example (vSphere problem):

[cuckoo.zip](https://github.com/cuckoosandbox/cuckoo/files/3846461/cuckoo.zip)

Why is that happening? Are there some configuration errors or maybe something else?

Thanks in advance!

**Comment by hiblackbear** [comment]
I had the same experience.
If you analyze 20 files using 20 virtual machines at the same time, the last 20th file requested will run after the first analysis is completed.

When process_results = no, 11 virtual machines work best.

**Comment by MetallDoktor** [comment]
In the end, the problem was resolved. We fully moved to the vSphere machinery and here is a thing: there is a large time delay due to full VM enumeration in the datacenter. We have about a thousand VM on single DC so we had to make some changes to the source code to get Cuckoo to enumerate VM in a specific directory. And that actually helped, analysis from file submission to report has become much faster.

Therefore I close this issue, hope it could help somebody else.
